workflow/snakemake_rules/remote_files.smk

```diff
@@ -0,0 +1,89 @@
+"""Shared functions to transparently handle local and remote inputs/outputs.
+"""
+from functools import partial
+from importlib import import_module
+from urllib.parse import urlsplit, urljoin
+
+
+def _remote_producer(init):
+    """
+    Produce a remote input function by loading a :class:`snakemake.remote`
+    module and passing the module to an *init* function.
+
+    The module is named by the first argument of the *init* function.  *init*
+    will be called with the loaded module and is expected to return a remote
+    input function suitable for use inside :func:`path_or_url`.
+
+    Returns a "producer" function which when called with no arguments returns
+    the remote input function.
+
+    The call to *init* (that is, the production of the remote input function)
+    is deferred until the first time the producer is called.
+
+    If the remote module is not available because required dependencies aren't
+    installed, then a :exc:`WorkflowError` exception object is returned on the
+    first call of the producer.
+    """
+    # Name of first argument to *init*.
+    remote_name = init.__code__.co_varnames[0]
+    remote = None
+
+    def producer():
+        nonlocal remote
+        if not remote:
+            remote = init(import_module(f"snakemake.remote.{remote_name}"))
```

Comment:
Right, the intention was to let Snakemake's WorkflowError be raised if optional deps weren't available, as those errors already describe the right deps to install. It's a good point though that they don't describe _how_ to install them, which might be tricky to do since "it depends" is the answer.

`boto3` gets pulled in as a transitive dep from several other deps already, both in our Conda environment and the Docker image. The Google deps are not in the Docker image, but they are in the Conda environment.

I thought this was a reasonable state to be in, but I can update the Docker image if desired.

I'll update the remote inputs docs to mention that optional deps may need to be installed for remote file support.

Reply:
Yeah, simply adding some docs should be enough here. I didn't realise it appears as a transitive dep in most situations, I guess that means I have a strange setup myself!

Reply:
Addressed in fc5c4583.